wfe/wfe.go

```diff
@@ -1489,22 +1498,50 @@ func (wfe *WebFrontEndImpl) FinalizeOrder(
 		return
 	}
 
+	// spliting order identifiers per types
+	var orderDNSs, orderIPs []string
+	for _, ident := range orderIdentifiers {
+		switch ident.Type {
+		case acme.IdentifierDNS:
+			orderDNSs = append(orderDNSs, ident.Value)
+		case acme.IdentifierIP:
+			orderIPs = append(orderIPs, ident.Value)
+		default:
+			wfe.sendError(acme.UnauthorizedProblem(
+				fmt.Sprintf("Order includes ilegal ident type %s", ident.Type)), response)
+		}
+	}
+	//and make uniqueLowerNames for DNSNames
+	orderDNSs = uniqueLowerNames(orderDNSs)
 	// Check that the CSR has the same number of names as the initial order contained
-	csrNames := uniqueLowerNames(parsedCSR.DNSNames)
-	if len(csrNames) != len(orderNames) {
+	csrDNSs := uniqueLowerNames(parsedCSR.DNSNames)
+	csrIPs := parsedCSR.IPAddresses
+	if len(csrDNSs) != len(orderDNSs) {
 		wfe.sendError(acme.UnauthorizedProblem(
-			"Order includes different number of names than CSR specifies"), response)
+			"Order includes different number of DNSnames than CSR specifies"), response)
+		return
+	}
+	if len(csrIPs) != len(orderIPs) {
+		wfe.sendError(acme.UnauthorizedProblem(
+			"Order includes different number of IP addresses than CSR specifies"), response)
 		return
 	}
 
 	// Check that the CSR's names match the order names exactly
-	for i, name := range orderNames {
-		if name != csrNames[i] {
+	for i, name := range orderDNSs {
+		if name != csrDNSs[i] {
 			wfe.sendError(acme.UnauthorizedProblem(
 				fmt.Sprintf("CSR is missing Order domain %q", name)), response)
 			return
 		}
 	}
+	for i, IP := range orderIPs {
+		if !csrIPs[i].Equal(net.ParseIP(IP)) {
+			wfe.sendError(acme.UnauthorizedProblem(
+				fmt.Sprintf("CSR is missing Order IP %q", IP)), response)
```

Comment:
I think the reason for me getting this error is that you don't sort the list of IP addresses. (`uniqueLowerNames()` sorts the list of DNS names, that's why the same loop for DNS names works fine.) Thus if there is more than one IP address in the list, and the order of the IPs in the CSR differs from the order of the IPs in the ACME order differ, you'll get this error.

Reply:
I was assumed that both csr domain name and order doesn't sorted, and thought order should've follow the SAN's order in csr to be vaild. add a sort at both order and csr IP list